### PR TITLE
[v23] vSphere VM Folder Fix

### PR DESCRIPTION
**Pull request by luthermonson**
https://github.com/rancher/rancher/issues/24840
The govmomi library was not recursively pulling vm folders so rewrote the folder lister to pull all folders and only show return results prefixed by `/<datacenter>/vm/` which shows an entire list now of all sub-folders.

e.g. 

```
{
"data": [
"",
"/RNCH-HE-FMT/vm/Discovered virtual machine",
"/RNCH-HE-FMT/vm/vm folder"
]
}
```

Will now be
```
{
"data": [
"",
"/RNCH-HE-FMT/vm/Discovered virtual machine",
"/RNCH-HE-FMT/vm/Discovered virtual machine/asdf",
"/RNCH-HE-FMT/vm/Discovered virtual machine/asdf/fdsa",
"/RNCH-HE-FMT/vm/vm folder",
"/RNCH-HE-FMT/vm/vm folder/level 1b",
"/RNCH-HE-FMT/vm/vm folder/level 1b/level 2bb",
"/RNCH-HE-FMT/vm/vm folder/level 1b/level 2ba",
"/RNCH-HE-FMT/vm/vm folder/level 1a",
"/RNCH-HE-FMT/vm/vm folder/level 1a/level 2aa",
"/RNCH-HE-FMT/vm/vm folder/level 1a/level 2aa/level 3aaa"
]
}
```

### CODE REVIEW

File: pkg/api/customization/vsphere/listers.go [hunk 1]
```diff
@@ -269,15 +270,18 @@ func listDataStoreClusters(ctx context.Context, finder *find.Finder) ([]string,
...
+	folders, err := finder.FolderList(ctx, "*")
 	if err != nil {
 		return nil, err
 	}
 
+	// base case of /<datacenter>/vm is covered by ""
 	data := []string{""}
 	for _, f := range folders {
-		data = append(data, f.InventoryPath)
+		if strings.HasPrefix(f.InventoryPath, fmt.Sprintf("%s/vm/", dc)) {
```
Comment (kinarashah) [inline pkg/api/customization/vsphere/listers.go]: nit: don't need to generate string `fmt.Sprintf("%s/vm/", dc)` every time, can move it outside loop

Comment (luthermonson) [reply pkg/api/customization/vsphere/listers.go]: great point, moving it.

File: pkg/api/customization/vsphere/listers.go [hunk 2]
```diff
@@ -269,15 +270,19 @@ func listDataStoreClusters(ctx context.Context, finder *find.Finder) ([]string,
...
 
-func listFolders(ctx context.Context, finder *find.Finder) ([]string, error) {
-	folders, err := finder.FolderList(ctx, "vm/*")
+func listFolders(ctx context.Context, finder *find.Finder, dc string) ([]string, error) {
+	folders, err := finder.FolderList(ctx, "*")
 	if err != nil {
 		return nil, err
 	}
 
+	prefix := fmt.Sprintf("%s/vm/", dc)
```
Comment (dramich) [inline pkg/api/customization/vsphere/listers.go]: Should use path.Join here

Comment (luthermonson) [reply pkg/api/customization/vsphere/listers.go]: sure

**Pull request merged**